lsp-positions/src/lib.rs

```diff
@@ -0,0 +1,353 @@
+// -*- coding: utf-8 -*-
+// ------------------------------------------------------------------------------------------------
+// Copyright © 2021, stack-graphs authors.
+// Licensed under either of Apache License, Version 2.0, or MIT license, at your option.
+// Please see the LICENSE-APACHE or LICENSE-MIT files in this distribution for license details.
+// ------------------------------------------------------------------------------------------------
+
+//! Defines LSP-compatible positioning information for source code.
+//!
+//! When writing a tool that analyzes or operates on source code, there's a good chance you need to
+//! interoperate with the [Language Server Protocol][lsp].  This seemingly simple requirement makes
+//! it surprisingly difficult to deal with _character locations_.  This is because Rust wants to
+//! store Unicode string content (i.e., the source code you're analyzing) in UTF-8, while
+//! LSP wants to specify character locations using [_UTF-16 code points_][lsp-utf16].
+//!
+//! That means that we ideally need to keep track of each source code position using at least two
+//! units: the UTF-8 offset within the file or containing line (to make it easy to index into UTF-8
+//! encoded strings), as well as the UTF-16 code point offset within the line (to make it possible
+//! to generate `Position` values for LSP).
```

Comment:
```suggestion
//! it surprisingly difficult to deal with _character locations_.  This is because Rust stores
//! Unicode string content (i.e., the source code you're analyzing) in UTF-8, while LSP specifies
//! character locations using [_UTF-16 code units_][lsp-utf16].
//!
//! For some background, Unicode characters, or code points, are encoded as one or more code units.
//! In UTF-8 a code unit is 1 byte, and a character is encoded in 1–4 code units (1–4 bytes).  In
//! UTF-16 a code unit is 2 bytes, and characters are encoded in 1–2 code units (2 or 4 bytes).
//! Rust strings are encoded as UTF-8, and indexed by byte (which is the same as by code unit).
//! Indices are only valid if they point to the first code unit of a code point.
//!
//! We keep track of each source code position using two units: the UTF-8 byte position within the
//! file or containing line, which can be used to index into UTF-8 encoded `str` and `[u8]` data,
//! and the UTF-16 code unit position within the line, which can be used to generate `Position`
//! values for LSP.
```

Reply:
This reads really well, thanks!  I added it to the crate README, as well.